### PR TITLE
Continue libovsdb migration

### PR DESCRIPTION
**Pull request by koolzz**
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

All changes must adhere to this template to make it easy for reviewers
and preserve rationale/history behind every change
-->

## 📑 Description
<!-- Add a brief description of the pr -->

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Migrates the node-side `ovs-vsctl` shell-outs to the libovsdb Go client. This reduces per-call subprocess fork/exec overhead, removes ad-hoc string parsing of `ovs-vsctl` output, and makes the OVS-side read/write paths consistent with how the rest of OVN-Kubernetes already talks to OVSDB.       

**This continues the work from(PR #6313), migration is incomplete and there is still work left to be done, but doing in chunks is easier**

                                                                                                                                                                                 
  ## Areas migrated
                                                                                                                                                                                                                                                   
  - `pkg/libovsdb/ops/`: new helpers — `DeletePortWithInterfaces`                                                                                                                                                                                  
    (bridge-scoped), `GetOVSInterface`, `CreateOrUpdatePodPort`; existing
    read wrappers gain `Equivalent: ovs-vsctl ...` docstrings so the                                                                                                                                                                               
    shell intent is still discoverable from code.                                                                                                                                                                                                  
  - `pkg/node`: `del-br`, `br-exists`, `del-port br-int`, port-to-br                                                                                                                                                                               
    lookups, and stale-pod-port cleanup all run through libovsdb.                                                                                                                                                                                  
    `ovsClient` is now plumbed through `BaseNodeNetworkController` so                                                                                                                                                                              
    every downstream caller has the same client.                                                                                                                                                                                                   
  - `pkg/util`, `cmd/ovn-kube-util`: `BridgeToNic`, `NicToBridge`, and                                                                                                                                                                             
    the `nicstobridge` read paths use libovsdb. `NicToBridge` waits for                                                                                                                                                                            
    the kernel netdev to materialize after the libovsdb transaction                                                                                                                                                                                
    because the daemon's link creation is asynchronous to the OVSDB                                                                                                                                                                                
    commit.                                                                                                                                                                                                                                        
  - `pkg/cni`: `ConfigureOVS` and `delRepPort` go through libovsdb when
    an `ovsClient` is available (privileged ovnkube-node path); the                                                                                                                                                                                
    unprivileged CNI shim path keeps the `ovs-vsctl` shell-out so the
    shim does not need libovsdb credentials.                                                                                                                                                                                                       
  - Tests: DPU and gateway tests now use the `libovsdbtest.NewOVSTestHarness`
    harness instead of mocking `ovs-vsctl` invocations.      

## Additional Information for reviewers
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [ ] My code requires changes to the documentation
- [ ] if so, I have updated the documentation as required
- [ ] My code requires tests
- [ ] if so, I have added and/or updated the tests as required
- [ ] All the tests have passed in the CI <!-- If not leave a comment as to why the CI is red and if you need help understanding what's wrong -->

## How to verify it
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Networking setup now uses a more robust OVSDB-backed flow for bridge, port, and interface handling.
  * CNI and node networking paths can now work with live OVS state for faster, more consistent configuration.

* **Bug Fixes**
  * Improved cleanup of stale bridges and ports during node and gateway operations.
  * Added safer handling for privileged CNI mode and better error reporting when OVS state is unavailable.
  * Updated tests to cover OVS-related edge cases and retry behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: go-controller/pkg/libovsdb/ops/vswitchdb.go [hunk 1]
```diff
@@ -73,6 +73,97 @@ func FindOVSPortsWithPredicate(ovsClient libovsdbclient.Client, p ovsPortPredica
...
+			return ops, nil
+		}
+		return nil, err
+	}
+
+	m := newModelClient(ovsClient)
+
+	// Cascade-delete each port's interfaces and the port itself. We hydrate
+	// each Port from cache via ovsClient.Get to learn its Interfaces UUIDs.
+	for _, portUUID := range bridge.Ports {
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/vswitchdb.go]: Are you sure this is required? My understanding is that OVS will garbage collect orphaned ports and interfaces once the bridge is removed.

File: go-controller/pkg/libovsdb/ops/vswitchdb.go [hunk 2]
```diff
@@ -237,9 +237,12 @@ func CreateOrUpdatePortWithInterfaceOps(ovsClient libovsdbclient.Client, ops []o
 	return m.CreateOrUpdateOps(ops, ifaceModel, portModel, bridgeModel)
 }
 
-// DeletePortWithInterfaces deletes an OVS port and all its interfaces from a bridge.
-// This removes both the Port and Interface objects, and detaches the port from the bridge.
-// This function is idempotent - it's safe to call even if the port doesn't exist.
+// DeletePortWithInterfaces deletes an OVS port and all its interfaces from a
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/vswitchdb.go:587]: What is `libovsdb/ops: bridge-scope DeletePortWithInterfaces` commit actually changing?
To my knowledge port/interface name has to be unique.

File: go-controller/pkg/libovsdb/ops/ovs/port.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+type portPredicate func(*vswitchd.Port) bool
+
+// GetOVSPort looks up an OVS port by name.
+func GetOVSPort(ovsClient libovsdbclient.Client, name string) (*vswitchd.Port, error) {
+	return libovsdbops.GetOVSPort(ovsClient, name)
+}
+
+// FindOVSPortsWithPredicate returns all OVS ports matching the predicate.
+func FindOVSPortsWithPredicate(ovsClient libovsdbclient.Client, p portPredicate) ([]*vswitchd.Port, error) {
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/ovs/port.go]: This re-implements `FindOVSPortsWithPredicate` from`go-controller/pkg/libovsdb/ops/vswitchdb.go`, why?

File: go-controller/pkg/libovsdb/ops/ovs/port.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	ctx, cancel := context.WithTimeout(context.Background(), types.OVSDBTimeout)
+	defer cancel()
+	var ports []*vswitchd.Port
+	err := ovsClient.WhereCache(p).List(ctx, &ports)
+	return ports, err
+}
+
+// CreateOrUpdatePortWithInterface creates or updates an OVS port and its
+// (single, type=internal) interface on a bridge in one transaction.
+func CreateOrUpdatePortWithInterface(ovsClient libovsdbclient.Client, bridgeName, portName string, portExternalIDs, ifaceExternalIDs map[string]string) error {
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/ovs/port.go]: `CreateOrUpdatePortWithInterface`, `CreateOrUpdatePortWithInterfaceOps`, `DeletePortWithInterfaces`, `DeletePortWithInterfacesOps` are pass-throughs to the same functions in vswitchdb.go. Can we keep them in one place? Having both is confusing.

File: go-controller/pkg/util/nicstobridge.go [hunk 5]
```diff
@@ -317,43 +325,48 @@ func BridgeToNic(bridge string) error {
...
+			}
+			if iface.Type != "patch" {
+				continue
+			}
+			peer := iface.Options["peer"]
+			if peer == "" {
+				klog.Warningf("Patch interface %q has no peer option", iface.Name)
+				continue
+			}
+			if err := ovsops.DeletePortWithInterfaces(ovsClient, "br-int", peer); err != nil {
```
Comment (kyrtapz) [inline go-controller/pkg/util/nicstobridge.go]: Have you considered removing all peers in one transaction? We could also include the bridge removal as well.

File: go-controller/pkg/node/default_node_network_controller.go [hunk 6]
```diff
@@ -560,8 +560,7 @@ func isOVNControllerReady() (bool, error) {
 	}
 
 	// check whether br-int exists on node
-	_, _, err = util.RunOVSVsctl("--", "br-exists", "br-int")
-	if err != nil {
+	if _, err := ovsops.GetBridge(ovsClient, "br-int"); err != nil {
```
Comment (kyrtapz) [inline go-controller/pkg/node/default_node_network_controller.go:565]: I know you cloned the existing  behavior but I think checking for `errors.Is(err, libovsdbclient.ErrNotFound)` as you did in https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6402/changes/151336c10ddbccc26aee30d5fa0cce1110758490 is correct and we should use it anywhere where we are replacing br-exists.

File: go-controller/pkg/node/gateway_init.go [hunk 7]
```diff
@@ -378,14 +378,14 @@ func (nc *DefaultNodeNetworkController) initGatewayMainStart(gw *gateway, waiter
 // and returns the name of the bridge if exists, or the interface itself
 // if the bridge needs to be created. In this last scenario, BridgeForInterface
 // will create the bridge.
-func interfaceForEXGW(intfName string) string {
-	if _, _, err := util.RunOVSVsctl("br-exists", intfName); err == nil {
+func interfaceForEXGW(ovsClient libovsdbclient.Client, intfName string) string {
+	if _, err := ovsops.GetBridge(ovsClient, intfName); err == nil {
```
Comment (kyrtapz) [inline go-controller/pkg/node/gateway_init.go:419]: Same as https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6402/changes/151336c10ddbccc26aee30d5fa0cce1110758490#r3426970189

File: go-controller/pkg/node/gateway_init.go [hunk 8]
```diff
@@ -378,14 +378,14 @@ func (nc *DefaultNodeNetworkController) initGatewayMainStart(gw *gateway, waiter
...
 		// It's a bridge
 		return intfName
 	}
 
 	bridge := util.GetBridgeName(intfName)
-	if _, _, err := util.RunOVSVsctl("br-exists", bridge); err == nil {
+	if _, err := ovsops.GetBridge(ovsClient, bridge); err == nil {
```
Comment (kyrtapz) [inline go-controller/pkg/node/gateway_init.go:427]: Same as Same as https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6402/changes/151336c10ddbccc26aee30d5fa0cce1110758490#r3426970189

File: go-controller/pkg/libovsdb/ops/vswitchdb.go [hunk 9]
```diff
@@ -73,6 +73,31 @@ func FindOVSPortsWithPredicate(ovsClient libovsdbclient.Client, p ovsPortPredica
 	return ports, err
 }
 
+// GetBridgeContainingPort returns the OVS bridge that owns the named port.
+// Returns ErrNotFound if the port does not exist or no bridge references it
+// in its ports column. This is the libovsdb equivalent of
+// `ovs-vsctl port-to-br <port>`.
+func GetBridgeContainingPort(ovsClient libovsdbclient.Client, portName string) (*vswitchd.Bridge, error) {
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/vswitchdb.go]: What would you think about adding something like `FindBridgeWithPredicate` and just calling it from here with the correct predicate?

File: go-controller/pkg/libovsdb/ops/ovs/port.go [hunk 10]
```diff
@@ -21,6 +21,13 @@ func GetOVSPort(ovsClient libovsdbclient.Client, name string) (*vswitchd.Port, e
 	return libovsdbops.GetOVSPort(ovsClient, name)
 }
 
+// GetBridgeContainingPort returns the OVS bridge that owns the named port,
+// the libovsdb equivalent of `ovs-vsctl port-to-br <port>`. Returns
+// ErrNotFound if the port does not exist or no bridge references it.
+func GetBridgeContainingPort(ovsClient libovsdbclient.Client, portName string) (*vswitchd.Bridge, error) {
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/ovs/port.go]: NIT: What do you think about calling it `GetPortBridge`?

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 11]
```diff
@@ -1944,16 +1945,23 @@ func newNodePortWatcher(
...
 			if err != nil {
-				return fmt.Errorf("failed to delete port %s stderr:%s (%v)", port, stderr, err)
+				if errors.Is(err, libovsdbclient.ErrNotFound) {
+					continue
+				}
+				return fmt.Errorf("failed to find bridge for port %s: %w", port.Name, err)
+			}
+			if err := ovsops.DeletePortWithInterfaces(ovsClient, bridge.Name, port.Name); err != nil {
+				return fmt.Errorf("failed to delete port %s: %w", port.Name, err)
 			}
```
[the remark this answers is not in the document]
Comment (kyrtapz) [reply go-controller/pkg/node/gateway_shared_intf.go:1969]: DeletePortWithInterfaces is not going to return `ErrNotFound`? RTFM

File: go-controller/pkg/libovsdb/ops/ovs/interface.go [hunk 12]
```diff
@@ -8,12 +8,18 @@ import (
...
 
+	libovsdbops "github.com/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/libovsdb/ops"
 	"github.com/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/types"
 	"github.com/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/vswitchd"
 )
 
 type interfacePredicate func(*vswitchd.Interface) bool
 
+// GetOVSInterface looks up an OVS interface by name.
+func GetOVSInterface(ovsClient libovsdbclient.Client, name string) (*vswitchd.Interface, error) {
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/ovs/interface.go]: What is the purpose of this pass-through function? Can we have one exposed `GetOVSInterface`?
Same applies to CreateOrUpdatePodPort/CreateOrUpdatePodPortOps.

File: go-controller/pkg/libovsdb/ops/ovs/port.go [hunk 13]
```diff
@@ -16,19 +16,17 @@ import (
...
-// GetOVSPort looks up an OVS port by name.
+// GetOVSPort: `ovs-vsctl find Port name=<name>`.
 func GetOVSPort(ovsClient libovsdbclient.Client, name string) (*vswitchd.Port, error) {
 	return libovsdbops.GetOVSPort(ovsClient, name)
 }
 
-// GetBridgeContainingPort returns the OVS bridge that owns the named port,
-// the libovsdb equivalent of `ovs-vsctl port-to-br <port>`. Returns
-// ErrNotFound if the port does not exist or no bridge references it.
+// GetBridgeContainingPort: `ovs-vsctl port-to-br <port>`.
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/ovs/port.go]: The previous, more verbose docstrings were more informative than the ovs-vsctl one-liners that replaced them.
I don't mind adding what `ovs-vsctl` it represents but I would keep the extra info we had.

This applies to all of the comments changed in d40fd6c4aa7e91cd42c74e12d33bffed51881040.

File: go-controller/pkg/libovsdb/ops/ovs/port.go [hunk 14]
```diff
@@ -16,19 +16,17 @@ import (
 
 type portPredicate func(*vswitchd.Port) bool
 
-// GetOVSPort looks up an OVS port by name.
+// GetOVSPort: `ovs-vsctl find Port name=<name>`.
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/ovs/port.go]: Can we keep the part where we say what the function does? I like the additional info though.

File: go-controller/pkg/libovsdb/ops/vswitchdb.go [hunk 15]
```diff
@@ -262,6 +280,82 @@ func CreateOrUpdatePortWithInterfaceOps(ovsClient libovsdbclient.Client, ops []o
...
+	if iface == nil {
+		return nil, fmt.Errorf("CreateOrUpdatePodPortOps: nil iface")
+	}
+	// Reject ports that already live on a different bridge. OVS's schema
+	// does not enforce one-bridge-per-port (Bridge.ports is a strong-set
+	// with no Port back-reference), so without this check the mutation
+	// below would leave the same Port UUID in two Bridges' ports lists.
+	// This matches the user-space safety check that `ovs-vsctl
+	// --may-exist add-port BRIDGE PORT` performs.
+	existing, err := GetBridgeContainingPort(ovsClient, portName)
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/vswitchdb.go]: NOTE: This check is still technically racy where we check the br and something attaches the port right after that but. I think this was the same previously and it is not really possible to happen in ovn-k so we should be good.

**Review by kyrtapz — COMMENTED** [review]
Thanks for the pr @koolzz, that's a great change overall!

I reviewed the first 7 commits(until d40fd6c4aa7e91cd42c74e12d33bffed51881040) and will continue on but wanted to post what I have so far. 
Please take a look at the CI failures if they can be related to the changes you are introducing.

File: go-controller/pkg/util/nicstobridge.go [hunk 16]
```diff
@@ -81,19 +97,14 @@ func GetNicName(brName string) (string, error) {
...
-	if stdout == "" && strings.HasPrefix(brName, "br") {
+	// Check for bridge-uplink to indicate the NIC.
+	uplink := br.ExternalIDs["bridge-uplink"]
+	if uplink == "" && strings.HasPrefix(brName, "br") {
 		// This would happen if the bridge was created before the bridge-uplink
 		// changes got integrated. Assuming naming format of "br<nic name>".
 		return brName[len("br"):], nil
 	}
-	return stdout, nil
+	return uplink, nil
```
[the remark this answers is not in the document]
Comment (kyrtapz) [reply go-controller/pkg/util/nicstobridge.go]: I think this is reasonable suggestion.

File: test/e2e/go.mod [hunk 17]
```diff
@@ -178,7 +186,10 @@ require (
 )
 
 replace (
+	// libovsdb's transitive deps want docker v28+, but
```
Comment (kyrtapz) [inline test/e2e/go.mod:189]: Can this change be a separate commit? This seems like an issue that should eventually be fixed in libovsdb.

File: go-controller/pkg/node/base_node_network_controller_dpu_test.go [hunk 18]
```diff
@@ -298,17 +319,24 @@ var _ = Describe("Node DPU tests", func() {
 					Cmd: genOVSFindCmd("30", "qos", "_uuid",
 						"external-ids:sandbox=a8d09931"),
 				})
-				// waitForPodInterface
+				// waitForPodInterface — still shell-out in the libovsdb path
```
Comment (kyrtapz) [inline go-controller/pkg/node/base_node_network_controller_dpu_test.go]: `waitForPodInterface` is calling ovs-vsctl every 200ms for every pod port(two in UDN) so this is one of the major CPU consumers when deploying a lot of pods at once. Why don't we just move to `ovsops.GetOVSInterface` and keep the polling? `ConfigureOVS` already has the client.

I think the `clearPodBandwidth` is also a low hanging fruit that would remove a bunch of unnecessary shell-outs. 

It would also be good to handle `deletePort` but it isn't wired for it today.

File: go-controller/pkg/util/nicstobridge.go [hunk 19]
```diff
@@ -33,50 +33,57 @@ func GetBridgeName(iface string) string {
...
+		err := ovsClient.Get(ctx, port)
+		cancel()
 		if err != nil {
-			return nil, fmt.Errorf("failed to get port %q on bridge %q:, stderr: %q, error: %v",
-				port, brName, stderr, err)
-
+			return nil, fmt.Errorf("failed to get port %s on bridge %q: %w", portUUID, brName, err)
+		}
+		ifaces := make([]*vswitchd.Interface, 0, len(port.Interfaces))
+		for _, ifaceUUID := range port.Interfaces {
```
Comment (kyrtapz) [inline go-controller/pkg/util/nicstobridge.go:66]: Can we use FindInterfacesWithPredicate instead of getting each interface?

File: go-controller/pkg/util/nicstobridge.go [hunk 20]
```diff
@@ -33,50 +33,57 @@ func GetBridgeName(iface string) string {
 	return fmt.Sprintf("br%s", iface)
 }
 
-// getBridgePortsInterfaces returns a mapping of bridge brName ports to its interfaces
-func getBridgePortsInterfaces(brName string) (map[string][]string, error) {
-	stdout, stderr, err := RunOVSVsctl("list-ports", brName)
+// getBridgePortsInterfaces returns a mapping of bridge brName ports to their
+// resolved Interface rows.
+func getBridgePortsInterfaces(ovsClient libovsdbclient.Client, brName string) (map[string][]*vswitchd.Interface, error) {
```
Comment (kyrtapz) [inline go-controller/pkg/util/nicstobridge.go]: NIT: This function could just take the bridge itself instead of just the name.

File: go-controller/pkg/util/nicstobridge.go [hunk 21]
```diff
@@ -33,50 +33,57 @@ func GetBridgeName(iface string) string {
...
+func GetNicName(ovsClient libovsdbclient.Client, brName string) (string, error) {
+	br, err := ovsops.GetBridge(ovsClient, brName)
+	if err != nil {
+		return "", fmt.Errorf("failed to get bridge %q: %w", brName, err)
+	}
+
 	// Check for system type port (required to be set if using NetworkManager)
-	var stdout, stderr string
-	portsToInterfaces, err := getBridgePortsInterfaces(brName)
+	portsToInterfaces, err := getBridgePortsInterfaces(ovsClient, brName)
```
Comment (kyrtapz) [inline go-controller/pkg/util/nicstobridge.go]: This isn't a problem introduced with your PR but this function seems more complex than the name suggests. I wonder if we could flip it so we first use `bridge-uplink` ext-id and only once that is empty we fallback to the logic iterating all interfaces.

File: go-controller/pkg/libovsdb/ops/vswitchdb.go [hunk 22]
```diff
@@ -98,6 +98,129 @@ func GetBridgeContainingPort(ovsClient libovsdbclient.Client, portName string) (
...
+				continue
+			}
+			return fmt.Errorf("failed to check existing bridge for port %q: %w", portName, err)
+		}
+		if existing.Name != bridgeName {
+			return fmt.Errorf("port %q is already attached to bridge %q", portName, existing.Name)
+		}
+	}
+
+	failMode := vswitchd.BridgeFailModeStandalone
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/vswitchdb.go]: NIT: `failMode` is only used once.

File: go-controller/pkg/libovsdb/ops/vswitchdb_test.go [hunk 23]
```diff
@@ -317,6 +317,40 @@ func TestCreateOrUpdatePodPort(t *testing.T) {
 	})
 }
 
+func TestCreateOrUpdateNicBridge(t *testing.T) {
+	t.Run("creates bridge with hwaddr, fail-mode, and bridge-id/bridge-uplink external_ids", func(t *testing.T) {
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/vswitchdb_test.go:452]: Can you add a test for updating the br?

File: LICENSES/packages/github.com/opencontainers/runc/v1.1.14/LICENSE [hunk 24]
```diff
@@ -1,191 +0,0 @@
-
```
Comment (kyrtapz) [inline LICENSES/packages/github.com/opencontainers/runc/v1.1.14/LICENSE:1]: What's up with all the license files removal?

File: go-controller/pkg/cni/helper_linux.go [hunk 25]
```diff
@@ -477,41 +482,190 @@ func setupSriovInterface(netns ns.NetNS, containerID, ifName string, ifInfo *Pod
...
-		return "", fmt.Errorf("failed to get ovn-pf-encap-ip-mapping, error: %v", err)
+func getPfEncapIP(ovsClient client.Client, deviceID string) (string, error) {
+	var mapping string
+	if ovsClient != nil {
+		ovs, err := ovsops.GetOpenvSwitch(ovsClient)
+		if err != nil {
+			return "", fmt.Errorf("failed to get Open_vSwitch row: %v", err)
+		}
+		mapping = ovs.ExternalIDs["ovn-pf-encap-ip-mapping"]
+	} else {
```
Comment (kyrtapz) [inline go-controller/pkg/cni/helper_linux.go:493]: I am not a fan of the implicit fallback to exec calls. My concern is that it will be very easy to miss that the privileged mode stopped using libovsdb if there is a bug somewhere in the future.

Ideally we should be able to build a light-weight client interface that could be used for both unprivileged and privileged modes.
I think that this would require some more work so the next best option in my view if would be to explicitly call the legacy path. I think the least we could do is to ensure that ovsClient is always set when `!config.UnprivilegedMode`. 

cc: @trozet if you have any better ideas.

**Review by kyrtapz — COMMENTED** [review]
Completed the rest of the commits, great work @koolzz!
Thank you for splitting the commits, it made reviewing the PR that much easier.

Please take a look at the failing tests.

**Comment by koolzz** [comment]
> Completed the rest of the commits, great work @koolzz! Thank you for splitting the commits, it made reviewing the PR that much easier.
> 
> Please take a look at the failing tests.

Hey thank you for the comprehensive review! I was on vacation so didn’t reply to the first set of comments, will address everything shortly

File: go-controller/pkg/libovsdb/ops/ovs/port.go (hunk 13 above)
Comment (koolzz) [reply go-controller/pkg/libovsdb/ops/ovs/port.go]: yeah I kept the original comment on the implementation function and shortened it on the pass through, although maybe I should just remove the passthrough stuff.

File: LICENSES/packages/github.com/opencontainers/runc/v1.1.14/LICENSE (hunk 24 above)
Comment (koolzz) [reply LICENSES/packages/github.com/opencontainers/runc/v1.1.14/LICENSE:1]: I've pulled it out into separate commit, this is just from running the `make -C go-controller third-party-licenses `

File: test/e2e/go.mod (hunk 17 above)
Comment (koolzz) [reply test/e2e/go.mod:189]: sure, https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6402/changes/cd219e53f7df627858d4fe0cb35f554177eed441

File: go-controller/pkg/libovsdb/ops/ovs/port.go (hunk 3 above)
Comment (koolzz) [reply go-controller/pkg/libovsdb/ops/ovs/port.go]: Collapsed it into vswitchdb.go

File: go-controller/pkg/util/nicstobridge.go (hunk 5 above)
Comment (koolzz) [reply go-controller/pkg/util/nicstobridge.go]: switched to bulk delete

File: go-controller/pkg/node/default_node_network_controller.go (hunk 6 above)
Comment (koolzz) [reply go-controller/pkg/node/default_node_network_controller.go:565]: added

File: go-controller/pkg/libovsdb/ops/vswitchdb.go (hunk 2 above)
Comment (koolzz) [reply go-controller/pkg/libovsdb/ops/vswitchdb.go:587]: its basically to make sure the port is only deleted if its on the provided bridge

File: go-controller/pkg/libovsdb/ops/ovs/interface.go (hunk 12 above)
Comment (koolzz) [reply go-controller/pkg/libovsdb/ops/ovs/interface.go]: Originally thought it would make the imports usage a bit cleaner but its not necessary so nuked it

File: go-controller/pkg/libovsdb/ops/ovs/port.go (hunk 4 above)
Comment (koolzz) [reply go-controller/pkg/libovsdb/ops/ovs/port.go]: nuked

File: go-controller/pkg/libovsdb/ops/vswitchdb.go (hunk 9 above)
Comment (koolzz) [reply go-controller/pkg/libovsdb/ops/vswitchdb.go]: we do have FindBridgesWithPredicate, that should suffice?

File: go-controller/pkg/cni/helper_linux.go (hunk 25 above)
Comment (koolzz) [reply go-controller/pkg/cni/helper_linux.go:493]: Yeah thats a good point I wasn't sure how to properly handle the privileged/unprivileged difference, I've added some checks to ensure ovsClient is set

Comment (koolzz) [reply go-controller/pkg/cni/helper_linux.go:493]: Should I leave the light-weight client interface as a followup or do you want it in this pr?

File: go-controller/pkg/node/base_node_network_controller_dpu_test.go (hunk 18 above)
Comment (koolzz) [reply go-controller/pkg/node/base_node_network_controller_dpu_test.go]: waitForPodInterface/clearPodBandwidth done, left deletePort for future cleanup. Please take a look

File: go-controller/pkg/util/nicstobridge.go (hunk 21 above)
Comment (koolzz) [reply go-controller/pkg/util/nicstobridge.go]: added short circuit for bridge-uplink check

**Comment by koolzz** [comment]
@kyrtapz implemented the requested changes, could you take another look?

File: go-controller/pkg/libovsdb/ops/vswitchdb.go [hunk 26]
```diff
@@ -146,9 +473,88 @@ func CreateOrUpdatePortWithInterfaceOps(ovsClient libovsdbclient.Client, ops []o
...
+		return nil, fmt.Errorf("port %q is already attached to bridge %q", portName, existing.Name)
+	}
+
+	iface.Name = portName
+	port.Name = portName
+	bridge := &vswitchd.Bridge{Name: bridgeName}
+
+	ifaceModel := operationModel{
+		Model:          iface,
+		OnModelUpdates: []interface{}{&iface.Type, &iface.Options, &iface.MTURequest, &iface.ExternalIDs},
```
Comment (kyrtapz) [inline go-controller/pkg/libovsdb/ops/vswitchdb.go]: When calling `CreateOrUpdatePodPort` when the interface already exists what will happen to the current external IDs? My understanding is that they will get overridden. 
That is a bit concerning because there are ext-ids that we do not control, like ovn-installed.

File: go-controller/pkg/libovsdb/ops/vswitchdb.go (hunk 2 above)
Comment (kyrtapz) [reply go-controller/pkg/libovsdb/ops/vswitchdb.go:587]: Good point. Would you mind adding a warning log when trying to delete a pod that is present but not on the provided bridge?

File: go-controller/pkg/cni/helper_linux.go (hunk 25 above)
Comment (kyrtapz) [reply go-controller/pkg/cni/helper_linux.go:493]: I think what we have now is sufficient and we can do a followup later.

File: go-controller/pkg/libovsdb/ops/ovs/port.go (hunk 4 above)
Comment (kyrtapz) [reply go-controller/pkg/libovsdb/ops/ovs/port.go]: I meant drop the pass-through wrappers, not delete ops/ovs/ :sweat_smile: 
I don't have very strong opinions around the file placement but I would like for us to avoid the refactor in this pr as it is already large.

**Review by kyrtapz — COMMENTED** [review]
Overall the PR looks good to me, just a few non-critical comments.

File: go-controller/pkg/libovsdb/ops/vswitchdb.go (hunk 26 above)
Comment (koolzz) [reply go-controller/pkg/libovsdb/ops/vswitchdb.go]: added optional removal https://github.com/ovn-kubernetes/ovn-kubernetes/compare/b192eedb457fd28aba6898f1749d74de4f6bcd8d..d71f68ef9a8d98cdfb7d0609e145f7c211b3f166#diff-01a2ef868473804cd3d9b894b093e5be0735cfb21bf27f375ece3a33cc07edc9R622, moved externalids to OnModelMutations.

**Comment by koolzz** [comment]
Implemented the requested changes from last review @kyrtapz, please take a look when you have time

File: go-controller/pkg/cni/helper_linux.go [hunk 27]
```diff
@@ -477,41 +482,192 @@ func setupSriovInterface(netns ns.NetNS, containerID, ifName string, ifInfo *Pod
...
+		if err := ovsops.CreateOrUpdatePodPort(ovsClient, "br-int", hostIfaceName, port, iface); err != nil {
+			return fmt.Errorf("failure in plugging pod interface: %v", err)
+		}
+		if stripDefaultNetIDs {
+			if err := ovsops.RemoveOVSInterfaceExternalIDs(ovsClient, hostIfaceName, types.NetworkExternalID, types.NADExternalID); err != nil {
+				return fmt.Errorf("failed to clear default-network external IDs from interface %s: %v", hostIfaceName, err)
+			}
+		}
+		return nil
+	}
```
Comment (kyrtapz) [inline go-controller/pkg/cni/helper_linux.go:633]: We can do a quick FUP if you prefer but why don't we do one transaction here? instead of two when `stripDefaultNetIDs` is true?

**Review by kyrtapz — DISMISSED** [review]

File: go-controller/pkg/cni/helper_linux.go (hunk 27 above)
Comment (koolzz) [reply go-controller/pkg/cni/helper_linux.go:633]: Good point, changed to ops model.

**Review by kyrtapz — APPROVED** [review]
